🌟 Exercise 1: Calculating Required Sample Size
You are planning an A/B test to evaluate the impact of a new email subject line on the open rate. Based on past data, you expect a small effect size of 0.3 (an increase from 20% to 23% in the open rate). You aim for an 80% chance (power = 0.8) of detecting this effect if it exists, with a 5% significance level (α = 0.05).

Calculate the required sample size per group using Python’s statsmodels library.
What sample size is needed for each group to ensure your test is properly powered?

In [3]:
import statsmodels.stats.power as power

effect_size = 0.3
alpha = 0.05
power_value = 0.8

sample_size = power.tt_ind_solve_power(
    effect_size=effect_size,
    alpha=alpha,
    power=power_value,
    alternative='two-sided'
)

print(f"Required sample size per group: {round(sample_size)}")

Required sample size per group: 175


# **We need 175 people in the original email group and 175 people in the new email group.**

🌟 Exercise 2: Understanding the Relationship Between Effect Size and Sample Size
Using the same A/B test setup as in Exercise 1, you want to explore how changing the expected effect size impacts the required sample size.

Calculate the required sample size for the following effect sizes: 0.2, 0.4, and 0.5, keeping the significance level and power the same.
How does the sample size change as the effect size increases? Explain why this happens.


In [4]:
effect_sizes = [0.2, 0.4, 0.5]
alpha = 0.05
power_value = 0.8

for effect_size in effect_sizes:
    sample_size = power.tt_ind_solve_power(
        effect_size=effect_size,
        alpha=alpha,
        power=power_value,
        alternative='two-sided'
    )
    print(f"Effect size: {effect_size}, Required sample size per group: {round(sample_size)}")

Effect size: 0.2, Required sample size per group: 393
Effect size: 0.4, Required sample size per group: 99
Effect size: 0.5, Required sample size per group: 64


# **Larger effect sizes (bigger improvements) require smaller sample sizes. This is because bigger changes are easier to detect. When we expect a smaller change, we need more data (a larger sample size) to be sure we're seeing a real effect.**

🌟 Exercise 3: Exploring the Impact of Statistical Power
Imagine you are conducting an A/B test where you expect a small effect size of 0.2. You initially plan for a power of 0.8 but wonder how increasing or decreasing the desired power level impacts the required sample size.

Calculate the required sample size for power levels of 0.7, 0.8, and 0.9, keeping the effect size at 0.2 and significance level at 0.05.
Question: How does the required sample size change with different levels of statistical power? Why is this understanding important when designing A/B tests?

In [5]:
effect_size = 0.2
alpha = 0.05
power_values = [0.7, 0.8, 0.9]

for power_value in power_values:
    sample_size = power.tt_ind_solve_power(
        effect_size=effect_size,
        alpha=alpha,
        power=power_value,
        alternative='two-sided'
    )
    print(f"Power: {power_value}, Required sample size per group: {round(sample_size)}")

Power: 0.7, Required sample size per group: 310
Power: 0.8, Required sample size per group: 393
Power: 0.9, Required sample size per group: 526


# **Higher power requires larger sample sizes. If you want to be very sure you'll see an effect if it exists, you need more data. Higher power reduces the chance of a false negative (or type 2 error).**

🌟 Exercise 4: Implementing Sequential Testing
You are running an A/B test on two versions of a product page to increase the purchase rate. You plan to monitor the results weekly and stop the test early if one version shows a significant improvement.

Define your stopping criteria.
Decide how you would implement sequential testing in this scenario.
At the end of week three, Version B has a p-value of 0.02. What would you do next?

Stopping Criteria:
A p-value below 0.05 (predefined significance level) for a certain period (like two weeks).
A maximum test duration.
Implementation:
Calculate p-values weekly.
Apply the stopping criteria after each week's data collection.
Adjust the significance level (alpha) to account for multiple looks (e.g., using the O'Brien-Fleming or Pocock boundaries).
Week Three Decision:
A p-value of 0.02 suggests a significant difference. However, it's crucial to consider:
Has the p-value been consistently low for multiple weeks?
Have you accounted for multiple looks?
If the p value is consistently low, and you have accounted for multiple look, then the test should be stopped, and version B should be implemented.

🌟 Exercise 5: Applying Bayesian A/B Testing
You’re testing a new feature in your app, and you want to use a Bayesian approach. Initially, you believe the new feature has a 50% chance of improving user engagement. After collecting data, your analysis suggests a 65% probability that the new feature is better.

Describe how you would set up your prior belief.
After collecting data, how does the updated belief (posterior distribution) influence your decision?
What would you do if the posterior probability was only 55%?

Prior Belief:
A 50% chance of improvement.
Posterior Distribution:
The 65% probability represents the posterior distribution. This updates the initial belief based on the data.
If the posterior is high enough, we would implement the new feature.
Posterior Probability of 55%:
A 55% probability indicates weak evidence.
In this scenario, we would likely continue the test to gather more data or revert to the original version.
Most companies set a higher threshold for implementing changes. For example they might want to see a 90 or 95% probability that the new feature is better.
This improvement is only a 5% shift from the prior 50% chance, it is a very small change, and more data would be needed.

🌟 Exercise 6: Implementing Adaptive Experimentation
You’re running a test with three different website layouts to increase user engagement. Initially, each layout gets 33% of the traffic. After the first week, Layout C shows higher engagement.

Explain how you would adjust the traffic allocation after the first week.
Describe how you would continue to adapt the experiment in the following weeks.
What challenges might you face with adaptive experimentation, and how would you address them?


Traffic Adjustment (Week 1):
Send more traffic to Layout C, less to A and B. How much more depends on how much better C performed.
Following Weeks:
Keep watching the results. If C stays best, keep giving it more traffic. If another layout improves, shift traffic again.
Challenges and Solutions:
Challenge: Balancing trying new layouts (Exploration: trying new things) vs. using the best one (Exploitation: Using what is already known to be effective).
Solution: Use algorithms to decide how much traffic each layout gets.
Challenge: It's more complicated to track and change traffic.
Solution: Use good tracking tools and set rules for traffic changes.